sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala

```diff
@@ -1076,4 +1076,49 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
         checkEvaluation(cast(ym, dt), value)
       }
   }
+
+  test("SPARK-35768: Take into account year-month interval fields in cast") {
+    Seq(("1-1", YearMonthIntervalType(YEAR, MONTH), 13),
+      ("-1-1", YearMonthIntervalType(YEAR, MONTH), -13))
+      .foreach { case (str, dataType, ym) =>
+        checkEvaluation(cast(Literal.create(str), dataType), ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL '$str' YEAR TO MONTH"), dataType), ym)
+        checkEvaluation(cast(Literal.create(s"INTERVAL -'$str' YEAR TO MONTH"), dataType), -ym)
+      }
+
+    Seq(("13", YearMonthIntervalType(YEAR), 156, 12),
+      ("13", YearMonthIntervalType(MONTH), 156, 13),
+      ("-13", YearMonthIntervalType(YEAR), -156, -12),
+      ("-13", YearMonthIntervalType(MONTH), -156, -13))
+      .foreach { case (str, dataType, year, month) =>
```

Comment:
nit:
```
Seq(("13", YearMonthIntervalType(YEAR), 13 * 12),
      ("13", YearMonthIntervalType(MONTH), 13),
      ("-13", YearMonthIntervalType(YEAR), -13 * 12),
      ("-13", YearMonthIntervalType(MONTH), -13))
      .foreach { case (str, dataType, ym) => {
  if (dataType == YearMonthIntervalType(YEAR)) {
          checkEvaluation(cast(Literal.create(s"INTERVAL '$str' YEAR"), dataType), year)
        } else {
          checkEvaluation(cast(Literal.create(s"INTERVAL '$str' MONTH"), dataType), month)
        }
}
```

Reply:
The value is always number of months, we don't need to check the year and month fields separately.